kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1398 @@
+type NetworkTopologySpec struct {
+	// The manual defined weights of the cluster
+	Weights WeightList `json:"weights,omitempty"`
+}
+
+// WeightList contains an array of WeightInfo objects.
+// +protobuf=true
+type WeightList []WeightInfo
+
+// CostList contains an array of OriginInfo objects.
+// +protobuf=true
+type CostList []OriginInfo
+
+// WeightInfo contains information about all network costs for a given algorithm.
+// +protobuf=true
+type WeightInfo struct {
+	// The name of the weightInfo (e.g., UserDefined)
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Costs between regions
+	RegionCostList CostList `json:"regionCostList" protobuf:"bytes,2,opt,name=regionCostList"`
+
+	// Costs between zones
+	ZoneCostList CostList `json:"zoneCostList" protobuf:"bytes,3,opt,name=zoneCostList"`
+}
+
+// OriginInfo contains information about network costs for a particular Origin.
+// +protobuf=true
+type OriginInfo struct {
+	// The name of the origin (e.g., Region Name, Zone Name).
+	Origin string `json:"origin" protobuf:"bytes,1,opt,name=origin"`
+
+	// Costs for the particular origin.
+	Costs []CostInfo `json:"destination" protobuf:"bytes,2,opt,name=destination"`
+}
+
+// CostInfo contains information about networkCosts.
+// +protobuf=true
+type CostInfo struct {
+	// Name of the destination (e.g., Region Name, Zone Name).
+	Destination string `json:"destination" protobuf:"bytes,1,opt,name=destination"`
+
+	// Bandwidth capacity between origin and destination. 
+	// +optional
+	BandwidthCapacity resource.Quantity `json:"bandwidthCapacity" protobuf:"bytes,2,opt,name=bandwidthCapacity"`
+
+	// Bandwidth allocatable between origin and destination.
+	// +optional
+	BandwidthAllocatable resource.Quantity `json:"bandwidthAllocatable" protobuf:"bytes,3,opt,name=bandwidthAllocatable"`
+
+	// Network Cost between origin and destination
+	NetworkCost int64 `json:"networkCost" protobuf:"bytes,4,opt,name=networkCost"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NetworkTopologyList is a collection of netTopologies.
+type NetworkTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of AppGroup
+	Items []NetworkTopology `json:"items"`
+}
+```
+
+## The inclusion of bandwidth in the scheduling process 
+
+### Bandwidth Requests via extended resources
+
+We plan to advertise the node's (physical) bandwidth capacity of the cluster via [extended resources](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/). 
+Bandwidth resources will be advertised as extended resources based on a bandwidth resource component open-sourced [here](https://github.com/jpedro1992/bandwidth-as-an-extended-resource). 
+These component will send HTTP requests to the Kubernetes API server regarding the node's bandwidth capacity. 
```

Comment:
How does the `bandwidth-as-an-extended-resource` component get the real bandwidth information from nodes?

Reply:
The plan is to extract the bandwidth capacity of the node from the [ethtool](https://linux.die.net/man/8/ethtool) command via its network interface (e.g., enp1s0f0).  

I will update the `bandwidth-as-an-extended-resource` repo in the next few days. We are currently working on the component. 

This is an example script to patch the nodes: 

```sh
#!/bin/sh
# Usage: ./script.sh enp1s0f0 http://10.2.0.113:8001

export INTERFACE="$1"
export API_URL="$2"

HOST=$(hostname)
URL="${API_URL}/api/v1/nodes/${HOST}/status"

echo "Network Bandwidth Component..."
echo $HOST
echo $INTERFACE
echo $API_URL
echo $URL

#speed=$(ethtool $INTERFACE | grep "Speed:" | sed 's/Speed://g')
#echo $speed

# Typically all values are in Mb/s
value=$(cat /sys/class/net/${INTERFACE}/speed)
speed=$(($value*1000000))
echo $speed

# Add the Patch
curl --header "Content-Type: application/json-patch+json" \
--request PATCH \
--data '[{"op": "add", "path": "/status/capacity/network.aware.com~1bandwidth", "value":'${speed}'}]' \
${URL}
```

Reply:
Will `bandwidth-as-an-extended-resource` be deployed as Daemonset? If so, does it mean the  `INTERFACE` option value will be the same in a cluster?(the start command for extended-resource pods needs to be exact the same in each daemonset pod)

Reply:
The idea is to deploy it as a Daemonset, but we only need to patch the node once. Then, the `network.aware.com/bandwidth` resources will be handled as CPU and Memory. 

The network interface can be different, it depends on the physical node. 

Instead of the .sh script, another option is to use the [ethtool package](https://pkg.go.dev/github.com/safchain/ethtool#section-readme) and build a GO program that patches the node as well.